File: r/R/hal9.R
.globals_nodes <- new.env()
.globals_data <- new.env()

Node <- R6::R6Class("Node", list(
    uid = NULL,
    fns = NULL,
    initialize = function(uid, fns) {
        self$uid <- uid
        self$fns <- fns
        register_node(self, self$uid)
        self
    },
    evaluate = function(fn, args) {
        fn <- self$fns[[fn]]

        if (is.null(fn)) {
            # stop("Function not defined for node.")
            return(NULL)
        }

        args <- lapply(args, h9_convert)

        result <- NULL
        if (length(args) == 0) {
            result <- fn()
        } else {
            result <- do.call(fn, args)
        }

        h9_convert(result)
    }
))

register_node <- function(obj, uid) {
    .globals_nodes[[uid]] <- obj
}

#' @export
h9_node <- function(uid, ...) {
    fns <- lapply(list(...), maybe_convert_to_fn)

    Node$new(uid, fns)
    invisible(NULL)
}

maybe_convert_to_fn <- function(x) {
    if (is.function(x)) {
        x
    } else {
        (\() x)
    }
}

#' @export
h9_set <- function(value, name) {
    .globals_data[[name]] <- value
    invisible(value)
}

#' @export
h9_get <- function(name) {
    .globals_data[[name]]
}

# returns a node obj
get_node <- function(uid) {
    .globals_nodes[[uid]]
}

process_request <- function(req) {
    responses <- lapply(req, function(call) {
        node <- get_node(call$node)
        fn_name <- call$fn_name

        if (is.null(node)) {
            return(
                list(
                    node = call$node,
                    fn_name = fn_name
                )
            )
        }

        fn_args_names <- call$args |> sapply(\(x) x$name)
        fn_args_values <- call$args |> lapply(\(x) x$value)
        fn_args <- setNames(fn_args_values, fn_args_names)

        tryCatch(
            {
                list(
                    node = call$node,
                    fn_name = fn_name,
                    result = node$evaluate(fn_name, fn_args)
                )
            },
            error = function(e) {
                list(
                    node = call$node,
                    fn_name = fn_name,
                    error = paste0(e, collapse = "\\n")
                )
            }
        )
    })
    list(calls = responses)
}

client_html <- function(...) {
    options <- list(...)
    options$designer <- list(
        persist = "pipeline",
        eval = "eval",
        heartbeat = "ping"
    )

    html <- paste(readLines(system.file("client.html", package = "hal9")), collapse = "\n")

    html <- gsub("__options__", gsub("'", "\\'", jsonlite::toJSON(options, auto_unbox = TRUE)), html)

    html
}

h9_run_script <- function(app, port = NULL) {
    tryCatch(
        {
            if (!file.exists(app)) {
                stop(paste0(app, "doesn't exist."))
            }

            source(app)

            app_path <- dirname(normalizePath(app))

            p <- plumber::pr() |>
                plumber::pr_get("/", function() {
                    client_html(mode = "run")
                }, serializer = plumber::serializer_html) |>
                plumber::pr_get("/design", function() {
                    client_html(mode = "design")
                }, serializer = plumber::serializer_html) |>
                plumber::pr_get("/pipeline", function() {
                    paste0(readLines(file.path(app_path, "")))
                }, serializer = plumber::serializer_text) |>
                plumber::pr_post("/pipeline", function(req) {
                    writeLines(req$postBody, file.path(app_path, "app.json"))
                    "{}"
                }, serializer = plumber::serializer_text) |>
                plumber::pr_post("/eval", function(req) {
                    parsed <- jsonlite::fromJSON(req$postBody, simplifyDataFrame = FALSE)
                    process_request(parsed)
                }, serializer = plumber::serializer_json(auto_unbox = TRUE, null = "null"))

            plumber::pr_run(p, port = port)
        },
        error = function(e) message("runtime_startup_error:", paste0(e, collapse = "\n"))
    )
}

File: py/src/hal9.py
import json
from typing import Callable, Any
import os
import sys

class _Node:
    """
    The base class which defines the backend execution graph
    """
    def __init__(self, uid:str =  None, funcs: dict[str, Callable] = None) -> None:
        self.uid = uid
        self.funcs = funcs
        _register_node(self)
    
    def evaluate(self, fn: str, *args, **kwargs):
        return self.funcs[fn](*args, **kwargs)

global_nodes:dict[str, _Node] = dict()
global_data:dict[str, Any] = dict()

def _register_node(node: _Node) -> None:
    global_nodes[node.uid] = node
    
def node(uid: str, **kwargs) -> None:
    _Node(uid, funcs=kwargs)


def get(x: str) -> _Node:
    if x in global_data.keys():
        return global_data[x]
    else:
        return None
                                                                                                                            
def set(name: str, value: Any) -> None:
    global_data[name] = value
    return value

    
def __process_request(calls: dict) -> None:
    response = dict()
    call_response = list()
    for call in calls:
        node = get(call['node'])
        kwargs = dict()
        for arg in call['args']:
            kwargs[arg['name']] = arg['value']
        result = node.evaluate(call['fn_name'], kwargs = kwargs)
        call_response.append({'node' : node.uid, 'fn_name': call['fn_name'], 'result' : result})
    response['calls'] = call_response
    return response

def __get_designer(**options: dict) -> str:
    options['designer'] = {
        'persist': 'pipeline',
        'eval': 'eval'
    }
    options = json.dumps(options)
    with open('../r/inst/client.html') as f:
        html = f.read()
    html = html.replace("__options__", options)
    return html

def start(path:str, port:int = 8000) -> None:
    from typing import Any, Dict, Union
    import uvicorn
    from fastapi import FastAPI
    import hal9 as h9
    import os
    from fastapi.responses import HTMLResponse, PlainTextResponse
    # from pydantic import BaseModel
    import webbrowser
    if not os.path.exists(path):
        os.makedirs(path)

    sys.path.append(path)

    if not os.path.exists(os.path.join(path, 'app.py')):
        with open(os.path.join(path, 'app.py'), 'w+') as fp:
            pass
    import app
    fastapp = FastAPI()

    # class Manifest(BaseModel):
    #     manifests: Dict[Any, Union[Any, None]]

    @fastapp.get('/pipeline', response_class=PlainTextResponse)
    async def get_pipeline():
        with open(os.path.join(path, 'app.json'), 'r') as f:
            return f.read()

    @fastapp.post('/pipeline', response_class=PlainTextResponse)
    async def write_pipeline(req:dict):
        with open(os.path.join(path, 'app.json'), 'w+') as f:
            f.write(json.dumps(req))
        return "{}"
    @fastapp.get('/', response_class=HTMLResponse)
    async def run_client():
        return h9.__get_designer(mode = "run")

    @fastapp.get('/design', response_class=HTMLResponse)
    async def get_designer():
        return h9.__get_designer(mode = "design")

    @fastapp.post("/eval")
    async def eval(manifest: dict):
        print(manifest)
        return h9.__process_request(manifest['calls'])
    a = webbrowser.open('http://127.0.0.1:'+str(port)+'/design')
    uvicorn.run(fastapp, host="127.0.0.1", port=port)
    